[LBSE] Fix performance of nested masks

rwlbuis opened the pull request:
#### 466a4e163671edd79710f625f32e3ee740b3a522
<pre>
[LBSE] Fix performance of nested masks
<a href="https://bugs.webkit.org/show_bug.cgi?id=271146">https://bugs.webkit.org/show_bug.cgi?id=271146</a>

Reviewed by Nikolas Zimmermann.

Masks with a high amount of nesting can take a lot of resources, both in computation and memory.
One example of this is mask-nested-reference.html (based on pattern-nested-reference.html).

To fix this, implement two optimizations:
- cache masks as image bitmaps (like legacy did)
- restrict the nesting level when calculating the mask bounding box
- take care of correctly invalidating in case mask contents change or a referencing client is added or removed

* LayoutTests/svg/custom/mask-nested-reference-expected.txt: Added.
* LayoutTests/svg/custom/mask-nested-reference.html: Added.
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.h:
(WebCore::RenderSVGResourceContainer::addReferencingCSSClient):
(WebCore::RenderSVGResourceContainer::removeReferencingCSSClient):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::applyMask):
(WebCore::RenderSVGResourceMasker::removeReferencingCSSClient):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.h:
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::adjustBoxForClippingAndEffects const):
* Source/WebCore/svg/SVGCircleElement.cpp:
(WebCore::SVGCircleElement::svgAttributeChanged):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::addReferencingCSSClient):
(WebCore::SVGElement::removeReferencingCSSClient):
* Source/WebCore/svg/SVGEllipseElement.cpp:
(WebCore::SVGEllipseElement::svgAttributeChanged):
* Source/WebCore/svg/SVGGraphicsElement.cpp:
(WebCore::SVGGraphicsElement::invalidateResourceImageBuffersIfNeeded):
* Source/WebCore/svg/SVGGraphicsElement.h:
* Source/WebCore/svg/SVGImageElement.cpp:
(WebCore::SVGImageElement::svgAttributeChanged):
* Source/WebCore/svg/SVGLineElement.cpp:
(WebCore::SVGLineElement::svgAttributeChanged):
* Source/WebCore/svg/SVGMaskElement.cpp:
(WebCore::SVGMaskElement::svgAttributeChanged):
(WebCore::SVGMaskElement::childrenChanged):
* Source/WebCore/svg/SVGPathElement.cpp:
(WebCore::SVGPathElement::svgAttributeChanged):
* Source/WebCore/svg/SVGPolyElement.cpp:
(WebCore::SVGPolyElement::svgAttributeChanged):
* Source/WebCore/svg/SVGRectElement.cpp:
(WebCore::SVGRectElement::svgAttributeChanged):
* Source/WebCore/svg/SVGResourceElementClient.h:
* Source/WebCore/svg/SVGTextContentElement.cpp:
(WebCore::SVGTextContentElement::svgAttributeChanged):
* Source/WebCore/svg/SVGTextElement.cpp:
(WebCore::SVGTextElement::childrenChanged):
* Source/WebCore/svg/SVGTextElement.h:
* Source/WebCore/svg/SVGTextPositioningElement.cpp:
(WebCore::SVGTextPositioningElement::svgAttributeChanged):
* Source/WebCore/svg/SVGUseElement.cpp:
(WebCore::SVGUseElement::svgAttributeChanged):

Canonical link: <a href="https://commits.webkit.org/277292@main">https://commits.webkit.org/277292@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a35b81f162dd73d221480dfed3ae80bd0cadf1d7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47111 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26286 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49750 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49794 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43160 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49418 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31558 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23747 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38376 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47692 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/23769 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40600 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19683 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21166 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41751 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5155 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43495 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42156 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51670 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22135 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18511 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45671 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23412 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/40746 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44676 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10415 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24194 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23129 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
```diff
@@ -268,8 +269,12 @@ void SVGBoundingBoxComputation::adjustBoxForClippingAndEffects(const SVGBounding
-            box.intersect(referencedMaskerRenderer->resourceBoundingBox(m_renderer, repaintRectCalculation));
+            static unsigned s_maskBoundingBoxNestingLevel = 0;
+            NestingLevelIncrementer incrementer { s_maskBoundingBoxNestingLevel };
+            if (s_maskBoundingBoxNestingLevel < 5) {
```
I don't understand this change. Are you trying to break cycles this way?

rwlbuis replied:
As discussed, this was a bit hacky, better to avoid any "inner" mask calculations, since they do not affect the "outer" mask.

nikolaszimmermann commented on Source/WebCore/svg/SVGGraphicsElement.cpp:
```diff
@@ -213,4 +215,16 @@ Path SVGGraphicsElement::toClipPath()
+        return;
+    if (CheckedPtr svgRenderer = dynamicDowncast<RenderLayerModelObject>(renderer())) {
+        if (auto* container = svgRenderer->layer()->enclosingSVGHiddenOrResourceContainer()) {
+            if (auto* patternRenderer = dynamicDowncast<RenderSVGResourceMasker>(container))
```
s/pattern/masker/

rwlbuis replied:
Ugh, missed it :) But I can fix once we land pattern PR.

nikolaszimmermann reviewed (approved):
Looks good, took some time to understand it.

nikolaszimmermann reviewed (approved):
Looks fine!

rwlbuis committed:
[LBSE] Fix performance of nested masks
https://bugs.webkit.org/show_bug.cgi?id=271146

Reviewed by Nikolas Zimmermann.

Masks with a high amount of nesting can take a lot of resources, both in computation and memory.
One example of this is mask-nested-reference.html (based on pattern-nested-reference.html).

To fix this, implement two optimizations:
- cache masks as image bitmaps (like legacy did)
- restrict the nesting level when calculating the mask bounding box
- take care of correctly invalidating in case mask contents change or a referencing client is added or removed

* LayoutTests/svg/custom/mask-nested-reference-expected.txt: Added.
* LayoutTests/svg/custom/mask-nested-reference.html: Added.
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.h:
(WebCore::RenderSVGResourceContainer::addReferencingCSSClient):
(WebCore::RenderSVGResourceContainer::removeReferencingCSSClient):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::applyMask):
(WebCore::RenderSVGResourceMasker::removeReferencingCSSClient):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.h:
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::adjustBoxForClippingAndEffects const):
* Source/WebCore/svg/SVGCircleElement.cpp:
(WebCore::SVGCircleElement::svgAttributeChanged):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::addReferencingCSSClient):
(WebCore::SVGElement::removeReferencingCSSClient):
* Source/WebCore/svg/SVGEllipseElement.cpp:
(WebCore::SVGEllipseElement::svgAttributeChanged):
* Source/WebCore/svg/SVGGraphicsElement.cpp:
(WebCore::SVGGraphicsElement::invalidateResourceImageBuffersIfNeeded):
* Source/WebCore/svg/SVGGraphicsElement.h:
* Source/WebCore/svg/SVGImageElement.cpp:
(WebCore::SVGImageElement::svgAttributeChanged):
* Source/WebCore/svg/SVGLineElement.cpp:
(WebCore::SVGLineElement::svgAttributeChanged):
* Source/WebCore/svg/SVGMaskElement.cpp:
(WebCore::SVGMaskElement::svgAttributeChanged):
(WebCore::SVGMaskElement::childrenChanged):
* Source/WebCore/svg/SVGPathElement.cpp:
(WebCore::SVGPathElement::svgAttributeChanged):
* Source/WebCore/svg/SVGPolyElement.cpp:
(WebCore::SVGPolyElement::svgAttributeChanged):
* Source/WebCore/svg/SVGRectElement.cpp:
(WebCore::SVGRectElement::svgAttributeChanged):
* Source/WebCore/svg/SVGResourceElementClient.h:
* Source/WebCore/svg/SVGTextContentElement.cpp:
(WebCore::SVGTextContentElement::svgAttributeChanged):
* Source/WebCore/svg/SVGTextElement.cpp:
(WebCore::SVGTextElement::childrenChanged):
* Source/WebCore/svg/SVGTextElement.h:
* Source/WebCore/svg/SVGTextPositioningElement.cpp:
(WebCore::SVGTextPositioningElement::svgAttributeChanged):
* Source/WebCore/svg/SVGUseElement.cpp:
(WebCore::SVGUseElement::svgAttributeChanged):

Canonical link: https://commits.webkit.org/277292@main
